Direct scanout doesn't work properly

### Regression?

Yes

### System Info and Version

<details> <summary>System/Version info</summary>

```sh
Hyprland, built from branch main at commit b9b8e6220f55af34e862b541a5a4b30ae6d8f15f  (renderer: fade out windows on silent moves).
Date: Thu Aug 29 00:54:49 2024
Tag: v0.42.0-86-gb9b8e622, commits: 5155

flags: (if any)


System Information:
System name: Linux
Node name: bek
Release: 6.10.6-zen1-1-zen
Version: #1 ZEN SMP PREEMPT_DYNAMIC Mon, 19 Aug 2024 17:02:05 +0000


GPU information: 
00:02.0 VGA compatible controller [0300]: Intel Corporation CometLake-U GT2 [UHD Graphics] [8086:9b41] (rev 02) (prog-if 00 [VGA controller])


os-release: NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://gitlab.archlinux.org/groups/archlinux/-/issues"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo


plugins:

======Config-Start======


Config File: /home/mastur/.config/hypr/hyprlandd.conf: Read Succeeded
monitor = eDP-1, 1920x1080, 0x0, 1

source = ~/.config/hypr/autostart.conf
source = ~/.config/hypr/vars.conf
source = ~/.config/hypr/binds.conf

env = QT_QPA_PLATFORMTHEME, qt6ct
env = HYPRCURSOR_THEME, rose-pine-cursor
env = HYPRCURSOR_SIZE, 24
env = XCURSOR_SIZE, 24

input {
    kb_layout = us,ru
    kb_options = grp:caps_toggle
    follow_mouse = 1
    force_no_accel = 1
    touchpad {
        natural_scroll = 1
	#disable_while_typing = 0
    }
    sensitivity = 0
}

general {
    gaps_in = 1 
    gaps_out = 2
    no_border_on_floating = 1
    border_size = 2
    col.active_border = rgba(03fc13ee) rgba(6b03fcee) 30deg
    col.inactive_border = rgba(595959aa)
    layout = dwindle
}

gestures {
    workspace_swipe = 1
}

decoration {
    rounding = 4
    blur {
        enabled = true
        size = 8
        passes = 2
        xray = 1
    }
    drop_shadow = no # Drop yourself from the window
    dim_inactive = yes
    dim_strength = 0.8
}

animations {
    enabled = yes
    ### BEZIERS ###
    bezier = eeee, 0.3, 0.2, 0.4, 0.9 # Govno = Poop
    bezier = aoaoa, 0.07, 0.65, 0.22, 0.98 # Huita = Bullshit
    ### BEZIERS ###
    animation = windows, 1, 3, aoaoa
    animation = windowsOut, 1, 5, default, popin 50%
    animation = border, 1, 10, default
    animation = borderangle, 1, 8, default
    animation = fade, 1, 7, default
    animation = workspaces, 1, 4, eeee
    first_launch_animation = false
}

dwindle {
    no_gaps_when_only = 1
    preserve_split = yes
}

misc {
    middle_click_paste = 0
    enable_swallow = 1
    swallow_regex = foot # Swallow my foot
#    vfr = 0 TESTING PURPOSE ONLY
}

xwayland {
    force_zero_scaling = 1 # XWayland rootless sucks
}

render {
#    explicit_sync = 0 # TESTING PURPOSE ONLY
    direct_scanout = 1
}

cursor {
    inactive_timeout = 5
}
debug {
    #overlay = 1 # TESTING PURPOSE ONLY
    #damage_blink = 1 # TESTING PURPOSE ONLY
    disable_logs = false
}

windowrulev2 = float, initialTitle:(vkcube) # Tiling vkcube = bruh
windowrule = fullscreen, ^(killingfloor-bin)$ # SDL1 - One Love
windowrule = nodim, ^(steam)$


Config File: /home/mastur/.config/hypr/autostart.conf: Read Succeeded
#exec-once = /usr/lib/xdg-desktop-portal-hyprland
exec-once = swww-daemon -f xrgb
exec-once = swww img /home/goose/Pictures/pstl.jpg
exec-once = waybar
exec-once = /usr/lib/polkit-kde-authentication-agent-1
exec-once = wl-paste --type text --watch cliphist store


Config File: /home/mastur/.config/hypr/vars.conf: Read Succeeded
$main = SUPER
$term = foot
$dmenu2 = bemenu-run -C -T --bottom
$dmenu = rofi -show drun
$lock = hyprlock
$screenshot = grimblast --freeze copy area
$clrpckr = pkill hyprpicker || hyprpicker -a


Config File: /home/mastur/.config/hypr/binds.conf: Read Succeeded
# Resize window from keyboard
bind=$main, R, submap, resize
submap=resize
binde=,right,resizeactive,10 0
binde=,left,resizeactive,-10 0
binde=,up,resizeactive,0 -10
binde=,down,resizeactive,0 10
bind=,catchall,submap,reset
submap=reset

bind = $main, T, exec, $term
bind = $main SHIFT, F, fullscreen
bind = $main SHIFT, Q, killactive, 
bind = $main SHIFT, M, exit, 
bind = $main, V, togglefloating, 
bind = $main, D, exec, $dmenu
bind = $main, S, exec, $dmenu2
bind = $main, J, togglesplit
bind = $main, L, exec, $lock
bind = $main, C, exec, $clrpckr
bind = $main, G, exec, ~/.config/hypr/gamemode.sh
bind = , Print, exec, $screenshot

# Focus move
bind = $main, left, movefocus, l
bind = $main, right, movefocus, r
bind = $main, up, movefocus, u
bind = $main, down, movefocus, d

# Window move
bind = $main SHIFT, left, swapwindow, l
bind = $main SHIFT, right, swapwindow, r
bind = $main SHIFT, up, swapwindow, u
bind = $main SHIFT, down, swapwindow, d

# Switch workspaces with win + [0-9]
bind = $main, 1, workspace, 1
bind = $main, 2, workspace, 2
bind = $main, 3, workspace, 3
bind = $main, 4, workspace, 4
bind = $main, 5, workspace, 5
bind = $main, 6, workspace, 6
bind = $main, 7, workspace, 7
bind = $main, 8, workspace, 8
bind = $main, 9, workspace, 9
bind = $main, 0, workspace, 10

# Move active window to a workspace with win + SHIFT + [0-9]
bind = $main SHIFT, 1, movetoworkspace, 1
bind = $main SHIFT, 2, movetoworkspace, 2
bind = $main SHIFT, 3, movetoworkspace, 3
bind = $main SHIFT, 4, movetoworkspace, 4
bind = $main SHIFT, 5, movetoworkspace, 5
bind = $main SHIFT, 6, movetoworkspace, 6
bind = $main SHIFT, 7, movetoworkspace, 7
bind = $main SHIFT, 8, movetoworkspace, 8
bind = $main SHIFT, 9, movetoworkspace, 9
bind = $main SHIFT, 0, movetoworkspace, 10

# Move/resize windows with win + LMB/RMB and dragging
bindm = $main, mouse:272, movewindow
bindm = $main, mouse:273, resizewindow

# THROWING FLASHBANG!!!!
bindl = , XF86MonBrightnessUp, exec, brightnessctl set +5%
bindl = , XF86MonBrightnessDown, exec, brightnessctl set 5%-

# Lock
bindl = , switch:Lid Switch, exec, $lock

======Config-End========
```

</details>

### Description

When i enter scanout mode, it doesn't render frames at all(it stucks) and Hyprland spams this in the log
> [LOG] CMonitorState::ensureBufferPresent: no buffer or mismatched format, attaching one from the swapchain for modeset being possible

Ping @ikalco 98e99cd

### How to reproduce

1) Make sure, that direct scanout is enabled
2) Enter scanout mode in apps, that requests other format in scanout, than the format, that Aquamarine/Hyprland uses(XR24 or XR30) eg. chromium related apps, vkcube-wayland 

### Crash reports, logs, images, videos

_No response_

``` /home/mastur/srcmk/Hyprland/src/helpers/Monitor.cpp: In member function ‘bool CMonitor::attemptDirectScanout()’:
/home/mastur/srcmk/Hyprland/src/helpers/Monitor.cpp:838:16: error: no match for ‘operator&&’ (operand types are ‘const Aquamarine::SDMABUFAttrs’ and ‘bool’)
  838 |     if (params && !params.success)
      |         ~~~~~~ ^~ ~~~~~~~~~~~~~~~
      |         |         |
      |         |         bool
      |         const Aquamarine::SDMABUFAttrs
/home/mastur/srcmk/Hyprland/src/helpers/Monitor.cpp:838:16: note: candidate: ‘operator&&(bool, bool)’ (built-in)
  838 |     if (params && !params.success)
      |         ~~~~~~~^~~~~~~~~~~~~~~~~~
/home/mastur/srcmk/Hyprland/src/helpers/Monitor.cpp:838:16: note:   no known conversion for argument 1 from ‘const Aquamarine::SDMABUFAttrs’ to ‘bool’
```
Compiler error

--- a/src/helpers/Monitor.cpp
+++ b/src/helpers/Monitor.cpp
@@ -832,6 +832,21 @@ bool CMonitor::attemptDirectScanout() {
 
     // FIXME: make sure the buffer actually follows the available scanout dmabuf formats
     // and comes from the appropriate device. This may implode on multi-gpu!!
+
+    const auto params = PSURFACE->current.buffer->buffer->dmabuf();
+    // scanout buffer isn't dmabuf, so no scanout
+    if (!params.success)
+        return false;
+
+    // entering into scanout, so save monitor format
+    if (lastScanout.expired())
+        prevDrmFormat = drmFormat;
+
+    if (drmFormat != params.format) {
+        output->state->setFormat(params.format);
+        drmFormat = params.format;
+    }
+
     output->state->setBuffer(PSURFACE->current.buffer->buffer.lock());
     output->state->setPresentationMode(tearingState.activelyTearing ? Aquamarine::eOutputPresentationMode::AQ_OUTPUT_PRESENTATION_IMMEDIATE :
                                                                       Aquamarine::eOutputPresentationMode::AQ_OUTPUT_PRESENTATION_VSYNC);

--- a/src/helpers/Monitor.hpp
+++ b/src/helpers/Monitor.hpp
@@ -100,7 +100,8 @@ class CMonitor {
     std::optional<Vector2D>     forceSize;
     SP<Aquamarine::SOutputMode> currentMode;
     SP<Aquamarine::CSwapchain>  cursorSwapchain;
-    uint32_t                    drmFormat = DRM_FORMAT_INVALID;
+    uint32_t                    drmFormat     = DRM_FORMAT_INVALID;
+    uint32_t                    prevDrmFormat = DRM_FORMAT_INVALID;
 
     bool                        dpmsStatus       = true;
     bool                        vrrActive        = false; // this can be TRUE even if VRR is not active in the case that this display does not support it.

--- a/src/render/Renderer.cpp
+++ b/src/render/Renderer.cpp
@@ -1277,7 +1277,8 @@ void CHyprRenderer::renderMonitor(CMonitor* pMonitor) {
             pMonitor->lastScanout.reset();
 
             // reset DRM format, make sure it's the one we want.
-            pMonitor->output->state->setFormat(pMonitor->drmFormat);
+            pMonitor->output->state->setFormat(pMonitor->prevDrmFormat);
+            pMonitor->drmFormat = pMonitor->prevDrmFormat;
         }
     }
 
@@ -1978,7 +1979,8 @@ bool CHyprRenderer::applyMonitorRule(CMonitor* pMonitor, SMonitorRule* pMonitorR
     pMonitor->currentMode   = nullptr;
 
     pMonitor->output->state->setFormat(DRM_FORMAT_XRGB8888);
-    pMonitor->drmFormat = DRM_FORMAT_XRGB8888;
+    pMonitor->prevDrmFormat = pMonitor->drmFormat;
+    pMonitor->drmFormat     = DRM_FORMAT_XRGB8888;
     pMonitor->output->state->resetExplicitFences();
 
     bool autoScale = false;
@@ -2219,7 +2221,8 @@ bool CHyprRenderer::applyMonitorRule(CMonitor* pMonitor, SMonitorRule* pMonitorR
 
     for (auto const& fmt : formats[(int)!RULE->enable10bit]) {
         pMonitor->output->state->setFormat(fmt.second);
-        pMonitor->drmFormat = fmt.second;
+        pMonitor->prevDrmFormat = pMonitor->drmFormat;
+        pMonitor->drmFormat     = fmt.second;
 
         if (!pMonitor->state.test()) {
             Debug::log(ERR, "output {} failed basic test on format {}", pMonitor->szName, fmt.first);

